File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.js
var lexicographicallySmallestArray = function(nums, limit) {
    const n = nums.length;

    const sorted = [...nums].sort((a, b) => a - b);

    // value -> group
    const numToGroup = new Map();

    // group -> values
    const groupToNums = new Map();

    let group = 0;

    numToGroup.set(sorted[0], group);
    groupToNums.set(group, [sorted[0]]);

    for (let i = 1; i < n; i++) {
        if (sorted[i] - sorted[i - 1] > limit) {
            group++;
        }

        numToGroup.set(sorted[i], group);

        if (!groupToNums.has(group)) {
            groupToNums.set(group, []);
        }

        groupToNums.get(group).push(sorted[i]);
    }

    // Pointer for each group
    const pointers = new Map();

    for (const g of groupToNums.keys()) {
        pointers.set(g, 0);
    }

    const result = new Array(n);

    for (let i = 0; i < n; i++) {
        const g = numToGroup.get(nums[i]);

        const index = pointers.get(g);

        result[i] = groupToNums.get(g)[index];

        pointers.set(g, index + 1);
    }

    return result;
};

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums, limit):
        n = len(nums)

        sorted_nums = sorted(nums)

        # value -> group
        num_to_group = {}

        # group -> list of values
        group_to_nums = {}

        group = 0

        num_to_group[sorted_nums[0]] = group
        group_to_nums[group] = [sorted_nums[0]]

        for i in range(1, n):
            if sorted_nums[i] - sorted_nums[i - 1] > limit:
                group += 1

            num_to_group[sorted_nums[i]] = group

            if group not in group_to_nums:
                group_to_nums[group] = []

            group_to_nums[group].append(sorted_nums[i])

        # Pointer for each group
        pointers = {g: 0 for g in group_to_nums}

        result = [0] * n

        for i in range(n):
            g = num_to_group[nums[i]]

            result[i] = group_to_nums[g][pointers[g]]
            pointers[g] += 1

        return result